src/main/java/net/neoforged/neoforge/fluids/FluidUnitSpec.java

```diff
@@ -0,0 +1,179 @@
+        FluidUnit defaultUnit = FluidUnits.BUCKET;
+        int smallestUnitDecimals = -1;
+
+        public Builder addUnit(FluidUnit unit, int value, boolean displayInTooltip) {
+            data.put(unit, new UnitInfo(value, displayInTooltip));
+            return this;
+        }
+
+        public Builder addUnit(FluidUnit unit, int value) {
+            data.put(unit, new UnitInfo(value));
+            return this;
+        }
+
+        public Builder defaultUnit(FluidUnit unit, int value) {
+            addUnit(unit, value, true);
+            defaultUnit = unit;
+            return this;
+        }
+
+        public Builder defaultUnit(FluidUnit unit) {
+            defaultUnit = unit;
+            return this;
+        }
+
+        public Builder smallestUnitDecimals(int decimals) {
+            smallestUnitDecimals = decimals;
+            return this;
+        }
+
+        public FluidUnitSpec build() {
+            return new FluidUnitSpec(checkFixedUnits(data).getOrThrow(), defaultUnit, smallestUnitDecimals);
+        }
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    private FluidUnitSpec(Map<FluidUnit, UnitInfo> data,
+            FluidUnit defaultUnit,
+            int smallestUnitDecimals) {
+        this.map = data;
+        this.listForTooltip = map.entrySet().stream()
+                .filter(e -> e.getValue().showInTooltip)
+                .sorted(Comparator.<Map.Entry<FluidUnit, UnitInfo>, Integer>comparing(e -> e.getValue().value).reversed())
+                .map(Map.Entry::getKey)
+                .toList();
+        this.defaultUnit = defaultUnit;
+        this.smallestUnitDecimals = smallestUnitDecimals;
+    }
+
+    /**
+     * Verify that if bucket or mb are present, they must have the correct values
+     */
+    private static DataResult<Map<FluidUnit, UnitInfo>> checkFixedUnits(Map<FluidUnit, UnitInfo> map) {
+        for (Map.Entry<FluidUnit, UnitInfo> entry : map.entrySet()) {
+            if (entry.getKey() == FluidUnits.BUCKET && entry.getValue().value != FluidType.BUCKET_VOLUME)
+                return DataResult.error(() -> "bucket must be defined as " + FluidType.BUCKET_VOLUME);
+            if (entry.getKey() == FluidUnits.MB && entry.getValue().value != 1)
+                return DataResult.error(() -> "mb must be defined as 1");
+        }
+        return DataResult.success(map);
+    }
+
+    /**
+     * Gets a list of all units in this spec..
+     */
+    public Set<FluidUnit> units() {
+        return map.keySet();
+    }
+
+    /**
+     * Gets a default unit that may be used by mods that want to display the quantity
+     * in a limited amount of space as a single number.
+     */
+    private FluidUnit defaultUnit() {
+        return this.defaultUnit;
+    }
+
+    /**
+     * Gets the quantity in millibuckets associated with a given unit.
+     */
+    public @Nullable Integer getValue(FluidUnit unit) {
```

Comment:
Rather than nullable, can we use some value that will never happen such as 0 or -1?

Reply:
I don't want the scenario of someone trying to get a unit that they're *certain* is there that isn't, and then blithely using the 0 or -1 returned.

Reply:
Another nitpick, this feels like it should be `OptionalInt` instead of a nullable `Integer` keeps the boxing and is much more clear that there either is some amount or something at all imo

Reply:
I have replaced it with Optional<Integer> in the latest version